README.md
MineFlags is a two player (hotseat or against the cpu) minesweeper game, similar to one that was very popular on the MSN Messenger app a few years ago.

It was made with Allegro 4.0, but you'll probably be able to compile with newer 4.x versions albeit with warnings.

The agup distribution may not refer to the original one, please check the agup's docs in the agup-0.11 directory.

Small font by Matthew Smith, big font by [iconian fonts](www.iconian.com).


agup-0.11/ChangeLog
2003-06-24  Peter Wang  <tjaden@users.sourceforge.net>

	* Released as version 0.11.

	* Fixed a couple of nits when compiling as C++.

	* agtk.c (d_agtk_list_proc): Disabled the ugly yellow border
	around the currently selected list item.

	* Incorporated a NeXTStep theme by Joao Neves.

2003-05-19  Peter Wang  <tjaden@users.sourceforge.net>

	* Released as version 0.10.

	* Incorporated Elias Pschernig's BeOS theme, with some minor
	changes for older gcc and Allegro.

2003-01-13  Peter Wang  <tjaden@users.sourceforge.net>

	* Released as version 0.9.2.

	* aphoton.c (photon_draw_menu_item): Bob fixed the background
	colour of the menu items.

2002-05-02  Peter Wang  <tjaden@users.sourceforge.net>

	* Released as version 0.9.1 (0.9 was a little early :-)

	* examp.c (lister): Small const mistake.

	* No longer #include aintern.h, but just copied the prototype for
	_draw_textbox into agupitrn.h.

2002-05-01  Peter Wang  <tjaden@users.sourceforge.net>

	* awin95.c (awin95_timer): Changed signature to be C++-friendly
	with `install_int'.

	* agtk.c: Changed instances of `inline' to `INLINE'.

	* Released version 0.9.

	* Made everything compilable without -Wno-unused parameter
	(caught a few not so good things).

	* Made everything compilable as C++ source (lots of damn
	typecasts, grr).

2002-03-30  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.8.

	* examp.c: Added vertical slider to example.

	* aphoton.c: Fixed vertical sliders.

2002-01-28  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.7.

2002-01-28  Eric Botcazou  <ebotcazou@multimania.com>

	* examp.c: Try 15-bit if 16-bit fails.

	* aphoton.c: Changed the look of the horizontal slider (to look
	like a "volume slider").  [pw: this should be done for the other
	themes as well]

2002-01-20  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.6.

2002-01-20  Eric Botcazou  <ebotcazou@multimania.com>

	* aphoton.c, aphoton.h: Added QNX Photon MicroGUI emulation.

2002-01-03  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.5.

2001-11-22  Peter Wang  <tjaden@users.sourceforge.net>

	* LICENCE.txt: Added zlib licence.

2001-11-18  Peter Wang  <tjaden@users.sourceforge.net>

	* examp.c: Formatting fix and use `void' for parmless functions.

2001-11-04  Peter Wang  <tjaden@users.sourceforge.net>

	* agtk.c, awin95.c: Replaced SEND_MESSAGE with object_message.

	* README: Wrote about d_agup_push_proc callback function.

	* d_*_push_proc: Redraw itself (as unpressed) before calling user
	callback, and pass the dialog entry that initiated the call.  Both
	were the old behaviour before David's implementation.

	* agtk.c, awin95.c: Gray text for disabled state in check and
	radio procs.

2001-10-27  Peter Wang  <tjaden@users.sourceforge.net>

	* README: Updated.

2001-10-21  David A. Capello  <dacap@users.sourceforge.net>

	* agtk.h, awin95.h: Added prototypes for window frame procs.
	
	* agtk.c (d_agtk_push_proc), awin95.c (d_awin95_push_proc):
	Cleaner implementation, changed callback function signature, and
	uses D_CLOSE internally instead of D_SELECTED.
	
	* aase.c, aase.h: Added a new widget set.

2001-10-13  Peter Wang  <tjaden@users.sourceforge.net>

	* agtk.c (d_agtk_button_proc): Gray text for disabled state.

	* examp.c: David Capello added examples of disabled and
	disabled/selected buttons.

	* awin95.c (d_awin95_shadow_box_proc): David Capello added this
	function which he forgot in the previous patch.

2001-10-07  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.4.

	* awin95.c: Reindented (David had already done some of this so I
	finished it off).

	* awin95.c: Changes from David A. Capello to buttons, lists and
	frames look more like Win95 widgets.

2001-10-04  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.3.

	* agtk.c (gtk_draw_scrollable_frame): David A. Capello fixed an
	aesthetic problem in scrollbars in list procs.

2001-09-16  Peter Wang  <tjaden@users.sourceforge.net>

	* The dp3 argument to d_agup_push_proc can now return messages.

2001-08-24  Peter Wang  <tjaden@users.sourceforge.net>

	* agup_*_proc renamed to d_agup_*_proc.

2001-08-08  Peter Wang  <tjaden@users.sourceforge.net>

	* To the user, themes are now constant opaque pointers.

	* agupitrn.h: Moved struct agup_thme into here from agup.h.

	* rules.mk: New file, containing AGUP rules.

	* Makefile: Changed a bit, preparing for guibrella.

2001-08-02  Peter Wang  <tjaden@users.sourceforge.net>

	* README: Updated for window procs.

	* agtk.c (agtk_window_proc): Very simple window proc for GTK.

	* examp.c (demo_windows): Added to show off agup_window_proc.

2001-07-31  Peter Wang  <tjaden@users.sourceforge.net>

	* awin95.c (awin95_window_proc): Adapted from DIME.
	* agup.c (agup_window_proc): Wrapper.

	* agtk.c (agtk_edit_proc), awin95.c (awin95_edit_proc): Use font
	height instead of DIALOG->h.

	* examp.c: Override gui_edit_proc as well, for the file selector.

2001-07-31  Sven Sandberg  <svsa1977@student.tdb.uu.se>

	* awin95.c (awin95_edit_proc): Win95 edit box shouldn't have a
	dotted rect around it.

2001-07-30  Peter Wang  <tjaden@users.sourceforge.net>

	* examp.c: Demonstrate file_select_ex.

	* examp.c: Use a select_theme function instead of select_<theme>.

2001-07-18  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.2.

	* README: Updated.

	* examp.c: Use themes.

	* agup.c: New module (theme manager).

2001-07-18  Robert J. Ohannessian  <ohannessian@videotron.ca>

	* awin95.c: New widget set.

2001-07-13  Peter Wang  <tjaden@users.sourceforge.net>

	* Released version 0.1.

2001-05-20  Peter Wang  <tjaden@users.sourceforge.net>

	* agtk.c: Added agtk_menu_proc and related.


agup-0.11/ans.h
/*
 * This file is part of the Allegro GUI Un-uglification Project.  If you
 * are looking for NeXTstep header files, you are looking in the wrong place.
 */

#ifndef _agup_included_ans_h
#define _agup_included_ans_h

extern int ans_fg_color;
extern int ans_bg_color;

void ans_init(void);
void ans_shutdown(void);

int d_ans_box_proc(int, DIALOG *, int);
int d_ans_shadow_box_proc(int, DIALOG *, int);
int d_ans_button_proc(int, DIALOG *, int);
int d_ans_push_proc(int, DIALOG *, int);
int d_ans_check_proc(int, DIALOG *, int);
int d_ans_radio_proc(int, DIALOG *, int);
int d_ans_icon_proc(int, DIALOG *, int);
int d_ans_edit_proc(int, DIALOG *, int);
int d_ans_list_proc(int, DIALOG *, int);
int d_ans_text_list_proc(int, DIALOG *, int);
int d_ans_textbox_proc(int, DIALOG *, int);
int d_ans_slider_proc(int, DIALOG *, int);
int d_ans_menu_proc(int, DIALOG *, int);
int d_ans_window_proc(int, DIALOG *, int);

extern AL_CONST struct agup_theme *ans_theme;

#endif


agup-0.11/abeos.h
/*
 * This file is part of the Allegro GUI Un-uglification Project.  If you
 * are looking for BeOS header files, you are looking in the wrong place.
 */

#ifndef _agup_included_abeos_h
#define _agup_included_abeos_h

extern int abeos_fg_color;
extern int abeos_bg_color;

void abeos_init(void);
void abeos_shutdown(void);

int d_abeos_box_proc(int, DIALOG *, int);
int d_abeos_shadow_box_proc(int, DIALOG *, int);
int d_abeos_button_proc(int, DIALOG *, int);
int d_abeos_push_proc(int, DIALOG *, int);
int d_abeos_check_proc(int, DIALOG *, int);
int d_abeos_radio_proc(int, DIALOG *, int);
int d_abeos_icon_proc(int, DIALOG *, int);
int d_abeos_edit_proc(int, DIALOG *, int);
int d_abeos_list_proc(int, DIALOG *, int);
int d_abeos_text_list_proc(int, DIALOG *, int);
int d_abeos_textbox_proc(int, DIALOG *, int);
int d_abeos_slider_proc(int, DIALOG *, int);
int d_abeos_menu_proc(int, DIALOG *, int);
int d_abeos_window_proc(int, DIALOG *, int);

extern AL_CONST struct agup_theme * abeos_theme;

#endif


agup-0.11/examp.c
/* examp.c
 * 
 * This file is part of the Allegro GUI Un-uglification Project.
 * It is a cheap demonstration of how pretty we are.
 * 
 * Peter Wang <tjaden@users.sourceforge.net>
 */


#include <allegro.h>
#include "agup.h"
#include "agtk.h"
#include "aphoton.h"
#include "awin95.h"
#include "aase.h"
#include "abeos.h"
#include "ans.h"


/*----------------------------------------------------------------------*/
/* Main dialog stuff							*/
/*----------------------------------------------------------------------*/


extern DIALOG main_dlg[];
extern MENU menu_theme[];


int select_theme(void)
{
    int c;
    agup_shutdown();
    agup_init((struct agup_theme *)active_menu->dp);

    for (c=0; menu_theme[c].proc; c++)
	menu_theme[c].flags &=~ D_SELECTED;
    active_menu->flags |= D_SELECTED;

    gui_fg_color = agup_fg_color;
    gui_bg_color = agup_bg_color;

    set_dialog_color(main_dlg, gui_fg_color, gui_bg_color);

    return D_REDRAW;
}


int file_selector(void)
{
    char path[1024] = "";
    file_select_ex("Testing...", path, NULL, sizeof path, 0, 0);
    return D_REDRAW;
}


int demo_windows(void)
{
    DIALOG dlg[] =
    {
	{ d_agup_window_proc, 0,   0,   320, 200, 0,             0,             0, 0, 0, 0, (void *)"A window",   NULL, NULL },
	{ d_ctext_proc,       160, 100, 0,   0,   agup_fg_color, agup_bg_color, 0, 0, 0, 0, (void *)"Press ESC.", NULL, NULL },
	{ NULL,               0,   0,   0,   0,   0,             0,             0, 0, 0, 0, NULL,                 NULL, NULL }
    };
    
    centre_dialog(dlg);
    do_dialog(dlg, -1);

    return D_REDRAW;
}


int surprise()
{
    alert("You pushed my button!", "Isn't this a useful feature?", 0,
	  "&Yes", "&Definitely", 'y', 'd');

    return D_O_K;
}


int quit_proc(int msg, DIALOG *d, int c)
{
    (void)d;
    (void)c;

    return (((msg == MSG_KEY) || (msg == MSG_XCHAR && c == KEY_F10 << 8))
	    && alert(0, "Quit?", 0, "Ok", "Cancel", 0, 0) == 1 ? D_CLOSE : D_O_K);
}


char *lister(int index, int *list_size)
{
    if (index >= 0) {
	return "hi";
    }
    else {
	index = 10;
	*list_size = index;
    }

    return NULL;
}


char sel[10];
char sel2[10];
char buf[101] = "101";
char buf2[101] = "dalmations";
char buf3[1024] = "textbox buffer, thaawfeebdaed kalsjd askjd xuuq"
		  "dassl!lhskld essdklsegassem lanimilbus zhjkasdh"
		  "kjash!! asdasd zxx,asd. ! XYZ";


MENU menu12[] =
{
    /* text		proc		child	flags		dp */
    { "Submenu",	0,		0,	0,		0 },
    { "",		0,		0,	0,		0 },
    { "Checked",	0,		0,	D_SELECTED,	0 },
    { "Disabled",	0,		0,	D_DISABLED,	0 },
    { NULL,		0,		0,	0,		0 },
};


MENU menu1[] =
{
    /* text		proc		child	flags	dp */
    { "&File selector",	file_selector,	0,	0,	0 },
    { "&Window frames", demo_windows,	0,	0,	0 },
    { "Test &1",	0,		0,	0,	0 },
    { "Test &2",	0,		menu12,	0,	0 },
    { "Test &3\thi",	0,		0,	0,	0 },
    { NULL,		0,		0,	0,	0 },
};


MENU menu_theme[] =
{
    /* text		proc		child	flags		dp */
    { "&GTK",		select_theme,	0,	D_SELECTED,	0 },
    { "&Photon",	select_theme,	0,	0,		0 },
    { "&Win95",		select_theme,	0,	0,		0 },
    { "&ASE",		select_theme,	0,	0,		0 },
    { "&BeOS",		select_theme,	0,	0,		0 },
    { "&NeXTStep",	select_theme,	0,	0,		0 },
    { NULL,		0,		0,	0,		0 },
};


MENU menubar[] =
{
    /* text		proc	child		flags	dp */
    { "&Menu 1",	0,	menu1,		0,	0 },
    { "&Theme",		0,	menu_theme,	0,	0 },
    { NULL,		0,	0,		0,	0 },
};


#define DIS D_DISABLED
#define SEL D_SELECTED


DIALOG main_dlg[] =
{
    /* proc                   x     y     w    h    fg    bg    key  flags    d1    d2  dp                    dp2    dp3 */
    { d_clear_proc,           0,    0,    0,   0,   0,    0,    0,   0,       0,    0,  0,                    0,     0 }, /* 0 */
    { d_button_proc,          1,    20,   98,  20,  0,    0,    0,   0,       0,    0,  (void *)"button",     0,     0 }, /* 1 */
    { d_agup_button_proc,     103,  20,   98,  20,  0,    0,    0,   0,       0,    0,  (void *)"button",     0,     0 }, /* 2 */
    { d_agup_push_proc,       206,  20,   98,  20,  0,    0,    0,   0,       0,    0,  (void *)"button",     0,     surprise }, /* 3 */
    { d_agup_button_proc,     309,  20,   98,  20,  0,    0,    0,   DIS,     0,    0,  (void *)"button",     0,     0 }, /* 4 */
    { d_agup_button_proc,     309,  45,   98,  20,  0,    0,    0,   DIS|SEL, 0,    0,  (void *)"button",     0,     0 }, /* 5 */
    { d_check_proc,           1,    45,   98,  20,  0,    0,    0,   0,       0,    0,  (void *)"check box",  0,     0 }, /* 6 */
    { d_agup_check_proc,      103,  45,   98,  20,  0,    0,    0,   0,       0,    0,  (void *)"check box",  0,     0 }, /* 7 */
    { d_radio_proc,           1,    69,   98,  20,  0,    0,    0,   0,       1,    1,  (void *)"radio 1",    0,     0 }, /* 8 */
    { d_agup_radio_proc,      103,  69,   98,  20,  0,    0,    0,   0,       1,    0,  (void *)"radio 2",    0,     0 }, /* 9 */
    { d_radio_proc,           1,    93,   98,  20,  0,    0,    0,   0,       1,    2,  (void *)"radio 3",    0,     0 }, /* 10 */
    { d_agup_radio_proc,      103,  93,   98,  20,  0,    0,    0,   0,       1,    0,  (void *)"radio 4",    0,     0 }, /* 11 */
    { d_shadow_box_proc,      1,    117,  98,  20,  0,    0,    0,   0,       0,    0,  0,                    0,     0 }, /* 12 */
    { d_agup_shadow_box_proc, 103,  117,  98,  20,  0,    0,    0,   0,       0,    0,  0,                    0,     0 }, /* 13 */
    { d_icon_proc,            206,  45,   48,  20,  0,    0,    0,   0,       2,    2,  0,                    0,     0 }, /* 14 */
    { d_agup_icon_proc,       256,  45,   48,  20,  0,    0,    0,   0,       2,    2,  0,                    0,     0 }, /* 15 */
    { d_edit_proc,            1,    156,  98,  14,  0,    0,    0,   0,       100,  0,  buf,                  0,     0 },
    { d_agup_edit_proc,       103,  154,  98,  14,  0,    0,    0,   0,       100,  0,  buf2,                 0,     0 },
    { d_text_list_proc,       1,    172,  98,  48,  0,    0,    0,   0,       2,    2,  lister,               sel,   0 },
    { d_agup_list_proc,       103,  172,  98,  48,  0,    0,    0,   0,       2,    2,  lister,               sel2,  0 },
    { d_agup_textbox_proc,    206,  70,   100, 100, 0,    0,    0,   0,       0,    0,  buf3,                 0,     0 },
    { d_slider_proc,          1,    140,  98,  12,  0,    0,    0,   0,       100,  0,  0,                    0,     0 }, /* horiz */
    { d_agup_slider_proc,     103,  140,  98,  12,  0,    0,    0,   0,       100,  0,  0,                    0,     0 },
    { d_slider_proc,          316,  70,   12,  98,  0,    0,    0,   0,       100,  0,  0,                    0,     0 }, /* vert */
    { d_agup_slider_proc,     336,  70,   12,  98,  0,    0,    0,   0,       100,  0,  0,                    0,     0 },
    { d_agup_menu_proc,       1,    1,    0,   0,   0,    0,    0,   0,       0,    0,  menubar,              0,     0 },
    { quit_proc,              0,    0,    0,   0,   0,    0,    27,  0,       0,    0,  0,                    0,     0 },
    { d_yield_proc,           0,    0,    0,   0,   0,    0,    0,   0,       0,    0,  0,                    0,     0 },
    { NULL,                   0,    0,    0,   0,   0,    0,    0,   0,       0,    0,  0,                    0,     0 }
};


void do_main_dialog()
{
    BITMAP *icon;

    agup_init(agtk_theme);

    gui_fg_color = agup_fg_color;
    gui_bg_color = agup_bg_color;
    gui_shadow_box_proc = d_agup_shadow_box_proc;
    gui_button_proc = d_agup_button_proc;
    gui_edit_proc = d_agup_edit_proc;
    gui_text_list_proc = d_agup_text_list_proc;

    menu_theme[0].dp = (void *)agtk_theme;
    menu_theme[1].dp = (void *)aphoton_theme;
    menu_theme[2].dp = (void *)awin95_theme;
    menu_theme[3].dp = (void *)aase_theme;
    menu_theme[4].dp = (void *)abeos_theme;
    menu_theme[5].dp = (void *)ans_theme;

    icon = load_bitmap("gem.pcx", 0);
    main_dlg[14].dp = icon;
    main_dlg[15].dp = icon;
    
    set_dialog_color(main_dlg, gui_fg_color, gui_bg_color);
    
    do_dialog(main_dlg, -1);

    destroy_bitmap(icon);
    agup_shutdown();
}


/*----------------------------------------------------------------------*/
/* Main  								*/
/*----------------------------------------------------------------------*/


int init()
{
    allegro_init();
    install_timer();
    install_keyboard();
    install_mouse();
    
    set_window_title("Widgets test");
  //  set_color_depth(16);
    if (set_gfx_mode(GFX_AUTODETECT, 320, 240, 0, 0) < 0) {
	set_color_depth(15);
	if (set_gfx_mode(GFX_AUTODETECT, 640, 480, 0, 0) < 0) {
	    allegro_message("Could not set graphics mode.\n");
	    return -1;
	}
    }
    
    return 0;
}


int main(void)
{
    if (init() < 0)
	return 1;
    show_mouse(screen);
    do_main_dialog();
    return 0;
}

END_OF_MAIN();
